sandbox: use sandboxService in CRI plugin instead of calling controller API directly

abel-von opened the pull request:
To make codes clean, CRI can store a Sandbox client in each sandbox in cache, and call the APIs in the client directly, so that we don't need to find the Controller everytime we need to call APIs of sandbox controller.

abel-von commented:
needs rebase after #9463 is merged

abel-von commented:
@fuweid @dmcgowan @mxpv

mikebrow commented on pkg/cri/server/sandbox_run.go:
```diff
@@ -400,21 +372,11 @@ func (c *criService) RunPodSandbox(ctx context.Context, r *runtime.RunPodSandbox
+	exitCh, err := sandboxClient.Wait(util.NamespacedContext())
+	if err != nil {
+		log.G(ctx).WithError(err).Error("failed to wait for sandbox exit")
+		return
```
return nil, err ?

abel-von replied:
done

mikebrow reviewed (commented):
see questions comments :-)

abel-von commented on pkg/cri/server/restart.go:
```diff
@@ -81,6 +82,27 @@ func (c *criService) recover(ctx context.Context) error {
 				return nil
 			}
+
+			// Convert the old pause Container to the new Sandbox
```
The comment here is updated now

mikebrow commented on pkg/cri/server/restart.go:
```diff
@@ -152,20 +175,21 @@ func (c *criService) recover(ctx context.Context) error {
 			}
-			exitCh <- *containerd.NewExitStatus(exit.ExitStatus, exit.ExitedAt, nil)
-		}()
+			// Sandbox was in running state, but its task has been deleted,
```
if it was unknown before restart... does that mean unknown is not valid after restart.. or are we removing unknown...

I'm probably missing something.

abel-von replied:
I think the status of sandbox is a little bit simpler than container, when we call `ListPodSandbox`, we can only get two states of whether `Ready` or `NotReady`. we only set the sandbox status to `Unknown` in cri cache temporary when we do `RunPodSandbox` and we do not store the state to db. So when containerd restart, it will check the status of sandbox by calling the `Status` or `Wait` APIs of sandbox controller. 
We can also see that in codes of  [recover.go](https://github.com/containerd/containerd/blob/main/pkg/cri/server/podsandbox/recover.go#L94-L114) and [restart.go](https://github.com/containerd/containerd/blob/main/pkg/cri/server/restart.go#L119-L137) that the status of sandbox is set to `Ready` and `NotReady`, without a third state.

mikebrow replied:
are you saying this:
https://github.com/containerd/containerd/blob/main/pkg/cri/server/podsandbox/sandbox_run.go#L285-L288](https://github.com/containerd/containerd/blob/main/pkg/cri/store/sandbox/status.go#L76-L78)

only applies to cache.. hmm is that new and we just didn't update the state diagram?  

Ok I missed the move from sandbox having a store that was actually stored to an in memory cache only..

hmm...

mikebrow replied:
missed this change: https://github.com/containerd/containerd/pull/7401

mikebrow replied:
Should I be concerned here:
https://github.com/containerd/containerd/blob/main/pkg/cri/server/podsandbox/sandbox_run.go#L265-L288
about not using a lock on this cached map (writing directly to the map contents to update state from unknown to ready, and other changes, without doing some sort of lock first via an update api?)

Guessing my confusion is more to the in between status of the cri sandbox store and it's meta information and the moving of that to the new sandbox meta...

abel-von replied:
> about not using a lock on this cached map (writing directly to the map contents to update state from unknown to ready, and other changes, without doing some sort of lock first via an update api?

I guess you are right, I reconstructed the code of PodSandbox about the status and add a commit in this PR, please take a look.

abel-von replied:
please take a look again. @mikebrow

dmcgowan commented:
> call the APIs in the client directly, so that we don't need to find the Controller everytime we need to call APIs of sandbox controller.

I'm hesitant on this one. We want to decouple the plugins from the `client` package rather than add more to client to cleanup plugins. It seems we are missing a service interface for controllers that should be in a plugin. The client container interface is not a good example to follow and also one we want to build a service interface around.

abel-von commented:
>  It seems we are missing a service interface for controllers that should be in a plugin. 

Removed the client dependency and add some methods in the `sandboxService` in cri. and fixed the bug that PinnedImages has no struct tag. please take a look.

abel-von commented:
/cc @dmcgowan

abel-von commented:
@dmcgowan @mxpv Do we still have to discuss this PR? as I did't use Sandbox Client anymore.

Burning1020 commented:
@mxpv Does it still need disscussion? We've disscussed at community meeting on Jan 25 before adding this label: https://docs.google.com/document/d/1Q8KyVJd26oAQ3MafbnkVBgJCopPl2Bw45H9L9br9Vus/edit#heading=h.3f10hgwx9mc

mxpv commented on internal/cri/server/sandbox_service.go:
```diff
@@ -17,31 +17,103 @@
+func (c *criSandboxService) SandboxStatus(ctx context.Context, sandboxer string, sandboxID string, verbose bool) (sandbox.ControllerStatus, error) {
+	sbController, ok := c.sandboxers[sandboxer]
+	if !ok {
+		return sandbox.ControllerStatus{}, fmt.Errorf("failed to get sandbox controller by %s", sandboxer)
```
I'd suggest introducing a helper so you don't have to replicate same error message in every function:

```go
func controller(name string) {
    sbController, ok := c.sandboxers[sandboxer]
 	if !ok {
 		return nil, fmt.Errorf("sandbox controller %q not found", sandboxer)
 	}

    return sbController
}
``

And then you could just return an error:

```go
controller, err := c.controller(sandboxer)
if err != nil {
    return nil, err
}
```

abel-von replied:
There is already an exported method named `SandboxController`, I was hesitant on calling it because we still need to check the error after calling, It seems we didn't reduce the code lines or complexity, So I gave up calling it.

abel-von replied:
done

mxpv reviewed (approved):
One minor suggestion, but overall look good to me once the CI is green.

fuweid commented on internal/cri/server/podsandbox/controller.go:
```diff
@@ -171,12 +171,16 @@ func (c *Controller) waitSandboxExit(ctx context.Context, p *types.PodSandbox, e
 		event := &eventtypes.TaskExit{ExitStatus: exitStatus, ExitedAt: protobuf.ToTimestamp(exitedAt)}
-		if cleanErr := handleSandboxTaskExit(dctx, p, event); cleanErr != nil {
+		if err := handleSandboxTaskExit(dctx, p, event); err != nil {
+			// TODO will backoff the event to the controller's own EventMonitor, but not cri's,
```
Not sure I understand your comment correctly: maybe we should let CRI plugin to retry all the failures? so that we don't need to implement other retry for shim-type sandbox controller.

abel-von replied:
`handleSandboxExit` in CRI plugin will only update the sandbox status in its `sandboxStore` and should ignore the implementation details of sandbox(So it should not know that if there is a `task` or `shim` for the sandbox), I think it is the podsandbox controller's responsibility to do cleanup of the legacy sandbox. and that is what we do in  `handleSandboxTaskExit`.

abel-von replied:
And the spliting of EventMonitor of podsandbox from CRI plugin is done in #9598, The retry of sandbox container task cleanup is done in that PR

fuweid reviewed (commented):
Leave some comments. Basically, It looks good.

abel-von commented on internal/cri/server/podsandbox/controller_test.go:
```diff
@@ -75,7 +74,10 @@ func Test_Status(t *testing.T) {
 	assert.Equal(t, s.State, sandboxstore.StateReady.String())
 
-	sb.Exit(*containerd.NewExitStatus(exitStatus, exitedAt, nil))
+	if err := sb.Exit(exitStatus, exitedAt); err != nil {
```
done

abel-von commented on internal/cri/server/sandbox_service.go:
```diff
@@ -17,31 +17,103 @@
+}
+
+func (c *criSandboxService) CreateSandbox(ctx context.Context, info sandbox.Sandbox, opts ...sandbox.CreateOpt) error {
+	sbController, ok := c.sandboxers[info.Sandboxer]
```
done

abel-von commented on internal/cri/server/service.go:
```diff
@@ -232,7 +239,7 @@ func NewCRIService(options *CRIServiceOptions) (CRIService, runtime.RuntimeServi
 
 	// Initialize pod sandbox controller
 	// TODO: Get this from options, NOT client
```
done

abel-von commented on internal/cri/server/podsandbox/types/podsandbox_test.go:
```diff
@@ -0,0 +1,71 @@
+	assert.Equal(t, p.Status.Get().State, sandbox.StateUnknown)
+	assert.Equal(t, p.ID, "test")
+	p.Metadata = sandbox.Metadata{ID: "test", NetNSPath: "/test"}
+	assert.Equal(t, p.Metadata.NetNSPath, "/test")
```
done

abel-von commented on internal/cri/server/podsandbox/types/podsandbox_test.go:
```diff
@@ -0,0 +1,71 @@
+		return status, nil
+	})
+	if err != nil {
+		t.Fatalf("Update pod sandbox status failed %v", err)
```
done

abel-von commented on internal/cri/server/podsandbox/types/podsandbox_test.go:
```diff
@@ -0,0 +1,71 @@
+		assert.Equal(t, exitTime, exitAt)
+	}()
+	time.Sleep(time.Second)
+	if err := p.Exit(uint32(128), exitAt); err != nil {
```
done
